### PR TITLE
Refactor reset mutable state API

### PR DESCRIPTION
**Pull request by wxing1292**
* Supporting insertion of transfer / timer / replication task.
* Add more check on last write version and workflow state

### CODE REVIEW

File: common/persistence/sql/sqlExecutionManager.go [hunk 1]
```diff
@@ -1880,23 +1885,92 @@ func createTimerTasks(
...
 }
 
-func assertAndUpdateCurrentExecution(tx sqldb.Tx, shardID int, domainID sqldb.UUID, workflowID string, newRunID sqldb.UUID, previousRunID sqldb.UUID,
+func assertNotCurrentExecution(tx sqldb.Tx, shardID int, domainID sqldb.UUID, workflowID string, runID sqldb.UUID) error {
+
+	assertFn := func(currentRow *sqldb.CurrentExecutionsRow) error {
+		if bytes.Equal(currentRow.RunID, runID) {
+			return &p.ConditionFailedError{Msg: fmt.Sprintf(
+				"Assertion on current record failed failed. Current run ID was %v, expected %v",
+				currentRow.RunID,
```
Comment (venkat1109) [inline common/persistence/sql/sqlExecutionManager.go]: this log is misleading - expected and got will always be same - I think you want to return error saying that the records are expected to be different but they are same

File: common/persistence/persistenceInterface.go [hunk 2]
```diff
@@ -368,19 +368,28 @@ type (
 
 	// InternalResetMutableStateRequest is used to reset workflow execution state  for Persistence Interface
 	InternalResetMutableStateRequest struct {
-		PrevRunID        string
+		// previous workflow information
+		PrevRunID            string
```
Comment (venkat1109) [inline common/persistence/persistenceInterface.go]: Does it make sense to create a separate persistence type for a "WorkflowRun" - that contains (runID, lastWriteVersion, state) ? Looks like these params are needed / passed everywhere for assertion.

File: common/persistence/sql/sqlExecutionManager.go [hunk 3]
```diff
@@ -1880,23 +1885,92 @@ func createTimerTasks(
 	return nil
 }
 
-func assertAndUpdateCurrentExecution(tx sqldb.Tx, shardID int, domainID sqldb.UUID, workflowID string, newRunID sqldb.UUID, previousRunID sqldb.UUID,
+func assertNotCurrentExecution(tx sqldb.Tx, shardID int, domainID sqldb.UUID, workflowID string, runID sqldb.UUID) error {
```
Comment (venkat1109) [inline common/persistence/sql/sqlExecutionManager.go]: I don't see this being used anywhere, do you need this ?

Comment (wxing1292) [reply common/persistence/sql/sqlExecutionManager.go]: this is from the 3+DC branch, probably leaked by rebase
i will remove this

File: common/persistence/persistenceInterface.go (hunk 2 above)
Comment (wxing1292) [reply common/persistence/persistenceInterface.go]: created an issue for this: #2013

**Pull request merged**